File: ch11/egypt.py
#  Simulate the Thebes, Tanis, Abydos example
from random import random

c, n = 0, 1_000_000
for i in range(n):
    r = random()
    if (r < 0.75):
        if (random() < 0.02): c += 1
    elif (r < (0.75+0.23)):
        if (random() < 0.03): c += 1
    else:
        if (random() < 0.10): c += 1

print("probability of priestly class = %0.5f" % (c/n))

File: ch11/egypt.c
#include <stdio.h>
#include <inttypes.h>
#include <time.h>

uint32_t seed;
uint32_t xorshift32(void) {
    seed ^= (seed<<13);
    seed ^= (seed>>17);
    seed ^= (seed<< 5);
    return seed;
}

double xorshift(void) {
    return xorshift32() / (double)0xFFFFFFFF;
}


int main(int argc, char *argv[]) {
    int i,c=0, n=1000000;
    double r;

    // seed xorshift
    seed = (uint32_t)time(NULL);

    // simulate
    for(i=0; i < n; i++) {
        r = xorshift();
        if (r < 0.75) {
            if (xorshift() < 0.02) c++;
        } else {
            if (r < 0.98) {
                if (xorshift() < 0.03) c++;
            } else {
                if (xorshift() < 0.10) c++;
            }
        }
    }

    printf("probability of priestly class = %0.5f\n", c/(double)n);

    return 0;
}
